lib/context_adapter_server.js

```diff
@@ -0,0 +1,427 @@
+            }
+          );
+        } else {
+          caLogger.error(
+            request.method.toUpperCase() + ' ' + request.url.path +
+            ', event=' + JSON.stringify(event), {
+              operationType: caConfig.OPERATION_TYPE.SERVER_LOG
+            }
+          );
+        }
+      }
+    });
+
+    server.connection({
+      host: host,
+      port: port
+    });
+
+    server.route([
+      {
+        method: 'GET',
+        path: '/version',
+        handler: function(request, reply) {
+          var message = caHelper.getVersion();
+          return reply(message);
+        }
+      },
+      {
+        method: 'POST',
+        path: caConfig.CA_PATH + '/updateContext',
+        handler: function(request, reply) {
+          var unicaCorrelatorPassed = request.headers[caConfig.UNICA_CORRELATOR_HEADER],
+              service = request.headers['fiware-service'],
+              servicePath = request.headers['fiware-servicepath'];
+
+          var meta = {
+            unicaCorrelator: unicaCorrelatorPassed || caHelper.getUnicaCorrelator(request),
+            transactionId: caHelper.getTransactionId(),
+            operationType: caHelper.getOperationType(request)
+          };
+
+          // Get the operation descriptor from the updateContext request received
+          caHelper.getOperationDescriptor(request, function(err, operationDescriptor) {
+            var ngsiResponse;
+
+            if (err) {
+              // The updateContext request is NOT well-formed
+              caLogger.warn(
+                'Invalid updateContext request received: ' +
+                request.method.toUpperCase() + ' ' + request.url.path +
+                ', error=' + err.code + ' - ' + err.message,
+                meta
+              );
+              // Respond to the received updateContext request
+              ngsiResponse = caHelper.getNGSIResponse(err, request);
+              return reply(ngsiResponse);
+            } else {
+              // The updateContext request is well-formed
+              caLogger.debug(
+                'Valid updateContext request received: ' +
+                request.method.toUpperCase() + ' ' + request.url.path +
+                ', payload=' + JSON.stringify(request.payload),
+                meta
+              );
+              caLogger.debug(
+                'operation descriptor: ' +
+                JSON.stringify(operationDescriptor),
+                meta
+              );
+              // Respond to the received updateContext request
+              ngsiResponse = caHelper.getNGSIResponse(null, request);
+              reply(ngsiResponse);
+              // Get the service description for the service id passed in the received request
+              caHelper.getServiceDescriptor(
+                service,
+                servicePath,
+                operationDescriptor.serviceId,
+                function(err, serviceDescriptor) {
```

Comment:
When you have multiple nested callbacks, as in this case, it's usually the occasion to start thinking about using composing libraries, such as [async.js](https://github.com/caolan/async).